### PR TITLE
Immortal solution to names

### LINKED ISSUE #542: Assign names after the introduction of variables.
I think I would prefer pushing names after the tactic. What it means is that
```
fresh n ih -> elim n,
[ zeroCase
, succCase
]
```
is replaced by something like
```
elim n,
[ zeroCase
, n ih => succCase
]
```
The names, `n` and `ih` were assigned just before the tactic. The use of `=>` has the extra bonus that
```
lam x => body
abs i => body
```
are really nothing special except there's no comma (and maybe scoping? not sure).

(EDIT: delete the `lam body` because `lam` actually takes a variant number of variables.)

### PR DESCRIPTION
**Pull request by jonsterling**
This is **_Jon-Carlo's Immortal Taste It Box for Arguments Unnamed Civilly (AUC)_**.

Resolves #542 

This is work in progress / proof of concept.

### CODE REVIEW

**Comment by favonia** [comment]
Hi, I am still hoping that each rule can reset the stack and we can omit `pop` in the syntax.

**Comment by jonsterling** [comment]
@favonia 

> Hi, I am still hoping that each rule can reset the stack and we can omit pop in the syntax.

Re: resetting the stack 'after each rule', I've tried to give some thoughts on that in the #542 thread; I'm not totally sure what is meant by omitting `pop` in the syntax. Did you mean that you just prefer to write `tac; [x => tac]` or something, rather than `tac; [pop x => tac]`? That certainly sounds plausible.

**Comment by favonia** [comment]
Yes, I just meant the syntactical omission, but with the current semantics of non-local stacks, I believe it is important to have `pop` to remind users of what is going on.

File: src/redprl/redprl.lex [hunk 1]
```diff
@@ -151,11 +151,10 @@ whitespace = [\ \t];
 <INITIAL>"claim"            => (Tokens.CLAIM (posTuple (size yytext)));
 <INITIAL>"use"              => (Tokens.USE (posTuple (size yytext)));
 <INITIAL>"with"             => (Tokens.WITH (posTuple (size yytext)));
+<INITIAL>"without"          => (Tokens.WITHOUT (posTuple (size yytext)));
```
Comment (cangiuli) [inline src/redprl/redprl.lex]: My Thought is so powerful!

**Comment by jonsterling** [comment]
Here's a cool example:
```
/* You can rename hypotheses using 'without' and 'with'; the scripts
 * have an appropriate binding structure, treating names pronominally.
 */
Thm RenamingHyps : [
  (-> bool int bool)
] by [
  lam x y => 
    without x y; 
    with a b => 
      use a
].
```

**Comment by favonia** [comment]
Hi, I wonder what the status of this PR is? I am interested in resolving #559 but that might create a conflict with this PR.

**Comment by jonsterling** [comment]
@favonia This PR still has some bugs in it, that Carlo and I are going to start fixing now that the LICS rush is over. We want to get it ready to merge very soon.

**Comment by jonsterling** [comment]
@favonia As for the motives, any merge conflicts that result from that will be easier to resolve, so I don't feel too strongly about what order this is done in.

**Comment by cangiuli** [comment]
@MatthewFluet You may be interested to hear about our latest MLton troubles (resolved by 5d2128d)!

8139c10 caused MLton to start taking multiple minutes to compile RedPRL (up from ~30 seconds), spent mostly in garbage collection during the `deepFlatten` pass. When we `-drop-pass deepFlatten` everything is back to normal, and this doesn't seem to affect either the binary's size or runtime. I'm attaching the `-verbose 3` output of compiling 8139c10 with and without `-drop-pass deepFlatten`.

[With deepFlatten: 64.55 + 172.43 (73% GC)](https://github.com/RedPRL/sml-redprl/files/1696661/log-flatten.txt) (`deepFlatten finished in 30.22 + 169.02 (85% GC)`)
[Without deepFlatten: 33.97 + 4.18 (11% GC)](https://github.com/RedPRL/sml-redprl/files/1696662/log-no-flatten.txt)

To be clear, everything's fine now that we drop the pass, but we thought you might be interested!

**Comment by jonsterling** [comment]
Going to merge, but let's leave the branch up for a while.

**Comment by MatthewFluet** [comment]
@cangiuli Thanks!  We've observed this problem in other programs (see the links at the bottom of http://mlton.org/DeepFlatten).

**Pull request merged**